src/ast/passes/unstable_feature.cpp

```diff
@@ -0,0 +1,83 @@
+#include <cstring>
+#include <string>
+
+#include "ast/ast.h"
+#include "ast/passes/unstable_feature.h"
+#include "ast/visitor.h"
+#include "bpftrace.h"
+#include "config.h"
+
+namespace bpftrace::ast {
+
+namespace {
+
+class UnstableFeature : public Visitor<UnstableFeature> {
+public:
+  explicit UnstableFeature(BPFtrace &bpftrace) : bpftrace_(bpftrace) {};
+
+  using Visitor<UnstableFeature>::visit;
+  void visit(AssignConfigVarStatement &assignment);
+  void visit(MapDeclStatement &decl);
+  void visit(Import &imp);
+
+private:
+  BPFtrace &bpftrace_;
+  std::unordered_set<std::string> enabled_unstable_features;
+};
+
+} // namespace
+
+void UnstableFeature::visit(AssignConfigVarStatement &assignment)
+{
+  // If set explicitly, we don't need to warn of use
+  if (bpftrace_.config_->is_unstable(assignment.var)) {
+    if (assignment.var == "unstable_map_decl" &&
+        bpftrace_.config_->unstable_map_decl) {
+      enabled_unstable_features.insert(assignment.var);
+    } else if (assignment.var == "unstable_import" &&
+               bpftrace_.config_->unstable_import) {
+      enabled_unstable_features.insert(assignment.var);
+    }
+  }
+}
+
+void UnstableFeature::visit(MapDeclStatement &decl)
+{
+  if (!bpftrace_.config_->unstable_map_decl) {
+    decl.addError() << "Map declarations are not enabled by default. To enable "
+                       "this unstable feature, set this config flag to 1 "
+                       "e.g. unstable_map_decl=1";
+    return;
+  }
+  if (!enabled_unstable_features.contains("unstable_map_decl")) {
+    decl.addWarning() << "Script is using an unstable feature. To prevent this "
```

Comment:
Did you want to just make the unstable_map_decl a new type to capture the TriState? (You could parse as `true`, `false`, `warn` and assign sensible mappings to 0 and 1 that match the current defaults. We'd effectively be changing the default from `false` to `warn`).

Reply:
If you do that you also might not need a dedicated pass. Individual passes would just need to check that config flag and see if it was e.g. UnstableFlag::Disable, UnstableFlag::Enable, UnstableFlag::Warn.

Reply:
What is the use-case for setting this config flag to `warn`? Couldn't users just not include the config flag at all and that's the default behavior?